node/src/main/scala/coop/rchain/node/Main.scala

```diff
@@ -167,15 +180,41 @@ object Main {
           log.error(s"Port $port is already in use!").as(false)
         )
 
-    List(
-      conf.server.port,
-      conf.server.kademliaPort,
-      conf.server.httpPort,
-      conf.grpcServer.portExternal,
-      conf.grpcServer.portInternal
-    ).traverse(isLocalPortAvailable)
-      .map(_.forall(identity))
-      .ifM(Task.unit, Task.delay(System.exit(1)))
+    def checkRChainProtocolPort(configuration: Configuration): Task[Configuration] =
+      isLocalPortAvailable(configuration.server.port).ifM(
+        configuration.pure[Task],
+        if (configuration.server.useRandomPorts)
+          for {
+            port <- getFreePort
+            _    <- log.info(s"Using random port $port as RChain Protocol port")
+          } yield configuration.copy(server = configuration.server.copy(port = port))
+        else
+          Task.delay(System.exit(1)).as(configuration)
```

Comment:
Is it ok exist just like this? Any resources to be cleaned up?

I would rather have a clean exit point from the main function instead of many exit points from these functions. Would that be difficult to achieve?

We should also print some error message.

Reply:
@sebtomba any specific reason why you decided to use `System.exit(1)` originally here?

Reply:
@itegulov I would create a seperate ticket in jira for that to get this cleaned up asap

Reply:
We have a growing number of `System.exit(1)` in the codebase. Maybe it's time to find a better approach in `Main`

Reply:
Ticketed: https://rchain.atlassian.net/browse/RCHAIN-3558